### DISCUSSION TITLE
reviewRecipients DonationVotingMerkleDistributionDirectTransferStrategy

### DISCUSSION THREAD

**Discussion by tnkshuuhei** [allo-protocol/allo-v2#457] [Q&A]
I'm trying to use DonationVoting strategy, and stuck on how to execute [reviewRecipients](https://github.com/allo-protocol/allo-v2/blob/14ef5206cfc5c9d68063b03abed2ee7c8564e73b/contracts/strategies/donation-voting-merkle-base/DonationVotingMerkleDistributionBaseStrategy.sol#L350) function.

https://github.com/allo-protocol/allo-v2/blob/14ef5206cfc5c9d68063b03abed2ee7c8564e73b/contracts/strategies/donation-voting-merkle-base/DonationVotingMerkleDistributionBaseStrategy.sol#L40-L58

I added some on foundry test, but only recipient[0] got accepted. what's wrong?
others return 0, the same things happened on [deployed strategy](https://sepolia.arbiscan.io/address/0x491d557fb2888403365ca68e354cf8dd35fcf907#events)
event was emitted but status not changed.
```
    function __register_accept_recipient() internal returns (address recipientId) {
        recipientId = __register_recipient();
        address recipientId2 = __register_recipient2();
        address recipientId3 = makeAddr("recipientId3");
        uint256 refRecipientsCounter = strategy.recipientsCounter();

        address[] memory recipientIds = new address[](3);
        recipientIds[0] = recipientId;
        recipientIds[1] = recipientId2;
        recipientIds[2] = recipientId3;

        DonationVotingMerkleDistributionBaseStrategy.ApplicationStatus[] memory statuses =
            new DonationVotingMerkleDistributionBaseStrategy.ApplicationStatus[](3);
        statuses[0] = __buildStatusRow(0, uint8(IStrategy.Status.Accepted));
        statuses[1] = __buildStatusRow(1, uint8(IStrategy.Status.Accepted));
        statuses[2] = __buildStatusRow(2, uint8(IStrategy.Status.Accepted));

        vm.prank(pool_admin());
        strategy.reviewRecipients(statuses, refRecipientsCounter);
        strategy.getRecipientStatus(recipientId);
        strategy.getRecipientStatus(recipientId2);
        strategy.getRecipientStatus(recipientId3);
    }

```
![Screenshot 2024-01-11 at 2 40 24](https://github.com/allo-protocol/allo-v2/assets/67859510/ae4aed85-262b-45a0-a67f-ec903b720d37)

**Comment by tnkshuuhei** [comment]
my questions are:
- what am I missing? (or wrong params)
- any good way to pass those params(ApplicationStatus[]) on frontend?

**Reply by codenamejason** [reply]
```solidity
recipientId = __register_recipient();
address recipientId2 = __register_recipient2();
address recipientId3 = makeAddr("recipientId3"); // not registered
```

Recipient 1 & 2 should be accepted and 3 not. I can't see your `__register_recipient2()` function to verify 2 .

**Reply by tnkshuuhei** [reply]
__register_recipient2() is here, this is prepared function. I didn't modify anything.
as you said recipient2 should be accepted.
https://github.com/allo-protocol/allo-v2/blob/14ef5206cfc5c9d68063b03abed2ee7c8564e73b/test/foundry/strategies/DonationVotingMerkleDistributionBase.t.sol#L875-L884

**Comment by 0xKurt** [comment]
1 and 2 should be accepted
and 3 not
```
recipientId = __register_recipient();
address recipientId2 = __register_recipient2();
address recipientId3 = makeAddr("recipientId3"); // not registered
```

**Reply by tnkshuuhei** [reply]
I answered here.
https://github.com/allo-protocol/allo-v2/discussions/457#discussioncomment-8085051

why this strategy is using bitmap instead of recipientid, status struct?
like this strategy is doing
https://github.com/allo-protocol/allo-v2/blob/14ef5206cfc5c9d68063b03abed2ee7c8564e73b/contracts/strategies/_poc/sqf-superfluid/SQFSuperFluidStrategy.sol#L419

**Reply by 0xKurt** [reply]
Tldr: using a bitmap makes the review of many recipients cheaper

**Reply by tnkshuuhei** [reply]
indeed, that makes sense.

@0xKurt @codenamejason 
btw I modified them on my local env, and passed all test (hardcoded in hurry).
I think this is more intuitive and easier to use. we can compare GAS.
https://github.com/tnkshuuhei/allo-v2/commit/8e42353879448f9192b34d4cf0ff76d8002c9c5c

**Reply by tnkshuuhei** [reply]
one more question, how bitmap is gonna be if we have more than 64 recipient?
